ntcore/src/main/native/cpp/local/LocalStorageImpl.cpp

```diff
@@ -0,0 +1,1086 @@
+                                           std::string_view logPrefix) {
+  auto datalogger = m_dataloggers.Add(m_inst, log, prefix, logPrefix);
+
+  // start logging any matching topics
+  auto now = nt::Now();
+  for (auto&& topic : m_topics) {
+    if (!PrefixMatch(topic->name, prefix, topic->special) ||
+        topic->type == NT_UNASSIGNED || topic->typeStr.empty()) {
+      continue;
+    }
+    topic->StartStopDataLog(datalogger, now, true);
+    // log current value, if any
+    if (topic->lastValue) {
+      topic->datalogs.back().Append(topic->lastValue);
+    }
+  }
+
+  return datalogger;
+}
+
+void StorageImpl::StopDataLog(NT_DataLogger logger) {
+  if (auto datalogger = m_dataloggers.Remove(logger)) {
+    // finish any active entries
+    auto now = Now();
+    for (auto&& topic : m_topics) {
+      topic->StartStopDataLog(datalogger.get(), now, false);
+    }
+  }
+}
+
+//
+// Schema functions
+//
+
+bool StorageImpl::HasSchema(std::string_view name) {
+  wpi::SmallString<128> fullName{"/.schema/"};
+  fullName += name;
+  auto it = m_schemas.find(fullName);
+  return it != m_schemas.end();
+}
+
+void StorageImpl::AddSchema(std::string_view name, std::string_view type,
+                            std::span<const uint8_t> schema) {
+  wpi::SmallString<128> fullName{"/.schema/"};
+  fullName += name;
+  auto& pubHandle = m_schemas[fullName];
+  if (pubHandle != 0) {
+    return;
+  }
+
+  auto topic = GetOrCreateTopic(fullName);
+
+  if (topic->localPublishers.size() >= kMaxPublishers) {
+    ERR("reached maximum number of publishers to '{}', not publishing",
+        topic->name);
+    return;
+  }
+
+  pubHandle = AddLocalPublisher(topic, {{"retained", true}},
+                                PubSubConfig{NT_RAW, type, {}})
+                  ->handle;
+
+  SetDefaultEntryValue(pubHandle, Value::MakeRaw(schema));
+}
+
+void StorageImpl::Reset() {
+  m_network = nullptr;
+  m_topics.clear();
+  m_publishers.clear();
+  m_subscribers.clear();
+  m_entries.clear();
+  m_multiSubscribers.clear();
+  m_dataloggers.clear();
+  m_nameTopics.clear();
+  m_listeners.clear();
+  m_topicPrefixListeners.clear();
+}
+
+void StorageImpl::NotifyTopic(LocalTopic* topic, unsigned int eventFlags) {
+  DEBUG4("NotifyTopic({}, {})", topic->name, eventFlags);
+  auto topicInfo = topic->GetTopicInfo();
+  if (!topic->listeners.empty()) {
+    m_listenerStorage.Notify(topic->listeners, eventFlags, topicInfo);
+  }
+
+  wpi::SmallVector<NT_Listener, 32> listeners;
+  for (auto listener : m_topicPrefixListeners) {
+    if (listener->multiSubscriber &&
+        listener->multiSubscriber->Matches(topic->name, topic->special)) {
+      listeners.emplace_back(listener->handle);
+    }
+  }
+  if (!listeners.empty()) {
+    m_listenerStorage.Notify(listeners, eventFlags, topicInfo);
+  }
+
+  if ((eventFlags & (NT_EVENT_PUBLISH | NT_EVENT_UNPUBLISH)) != 0) {
+    if (!m_dataloggers.empty()) {
+      auto now = Now();
```

Comment:
Just noting that caching `Now()` wasn't done before. I don't think it matters, but for posterity (on the off chance it does matter), this was added in [[ntcore] Split LocalStorage implementation into separate files](https://github.com/wpilibsuite/allwpilib/pull/7216/commits/63e71a285125e35723998fe57883ee0f5d467eff).

Reply:
Yep, no harm to cache it, and it's actually a tiny performance improvement.